subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/action/TestExecutionRequestAction.java

```diff
@@ -69,14 +73,42 @@ public TestExecutionRequestAction(BuildEventSubscriptions clientSubscriptions,
     // TestExecutionRequestAction
     public static TestExecutionRequestAction create(BuildEventSubscriptions clientSubscriptions, StartParameterInternal startParameter, ProviderInternalTestExecutionRequest testExecutionRequest) {
         ImmutableSet<String> classNames = ImmutableSet.copyOf(testExecutionRequest.getTestClassNames());
-        return new TestExecutionRequestAction(clientSubscriptions, startParameter,
+        List<InternalTaskSpec> taskSpecs = testExecutionRequest.getTaskSpecs(Collections.emptyList());
+        boolean runDefaultTasks = testExecutionRequest.isRunDefaultTasks(false);
+
+        return new TestExecutionRequestAction(
+            clientSubscriptions,
+            configureStartParameter(startParameter, taskSpecs, runDefaultTasks),
             ImmutableSet.copyOf(testExecutionRequest.getTestExecutionDescriptors()),
             classNames,
             getInternalJvmTestRequests(testExecutionRequest, classNames),
             getDebugOptions(testExecutionRequest),
             getTaskAndTests(testExecutionRequest),
-            testExecutionRequest.isRunDefaultTasks(false),
-            testExecutionRequest.getTaskSpecs(Collections.emptyList()));
+            runDefaultTasks,
+            taskSpecs);
+    }
+
+    private static StartParameterInternal configureStartParameter(StartParameterInternal startParameter, List<InternalTaskSpec> taskSpecs, boolean runDefaultTasks) {
+        Preconditions.checkArgument(
+            startParameter.getTaskNames().isEmpty(),
+            "Cannot pass task requests with start parameter here, got %s",
+            startParameter.getTaskNames());
+        if (!taskSpecs.isEmpty()) {
+            List<String> taskPaths = taskSpecs.stream().map(InternalTaskSpec::getTaskPath).collect(Collectors.toList());
+
+            Preconditions.checkArgument(
+                !runDefaultTasks,
+                "Cannot run default tasks when task specs %s are provided",
+                taskPaths);
+
+            startParameter.setTaskRequests(Collections.singletonList(new DefaultTaskExecutionRequest(taskPaths)));
```

Comment:
💅 I'd move this side-effect to the caller and make this method simply return the new list of requests.